internal/heartbeat/heartbeat_test.go

```diff
@@ -0,0 +1,126 @@
+package heartbeat_test
+
+import (
+	"context"
+
+	"github.com/golang/mock/gomock"
+	"github.com/jakub-dzon/k4e-device-worker/internal/configuration"
+	"github.com/jakub-dzon/k4e-device-worker/internal/datatransfer"
+	"github.com/jakub-dzon/k4e-device-worker/internal/hardware"
+	"github.com/jakub-dzon/k4e-device-worker/internal/heartbeat"
+	"github.com/jakub-dzon/k4e-device-worker/internal/workload"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"google.golang.org/grpc"
+
+	pb "github.com/redhatinsights/yggdrasil/protocol"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Heartbeat", func() {
+
+	var (
+		datadir       = "/tmp"
+		mockCtrl      *gomock.Controller
+		wkManager     *workload.WorkloadManager
+		configManager *configuration.Manager
+		hw            = &hardware.Hardware{}
+		monitor       = &datatransfer.Monitor{}
+		hb            = &heartbeat.Heartbeat{}
+		err           error
+		client        DispatcherInstance
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		wkwMock := workload.NewMockWorkloadWrapper(mockCtrl)
+
+		wkwMock.EXPECT().Init().Return(nil).AnyTimes()
+		wkwMock.EXPECT().List().AnyTimes()
+		wkwMock.EXPECT().PersistConfiguration().AnyTimes()
+
+		wkManager, err = workload.NewWorkloadManagerWithParams(datadir, wkwMock)
```

Comment:
At this point, it is the best way that I found. TBH, no reason to mock the direct collaborator, mainly because is all our code. Meanwhile, WorkloadWrapper is a wrapper on top of Podman, that we need to mock to avoid using 3party tooling, no?

Reply:
Yes and no:) Mocking your own code is a way to ensure that you're testing one thing and one thing only. When you mock collaborator's collaborator you test both code under test and the direct collaborator. And when the tests fail you need to dig into both code you're working on and collaborator (which should be covered by its own set of tests).